Add tags as labels to prometheus metrics

proffalken opened the pull request:
# Description

Fixes #680 - Once working, this should add any tags added in the UI as labels for Prometheus

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and test it
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

Please do not use any external image service. Instead, just paste in or drag and drop the image here and it will be uploaded automatically.

proffalken commented:
@chakflying / @louislam - I'm not sure how to do this in NodeJS, hopefully this is a reasonable start to show the kind of thing I'm trying to achieve?

Instead of 

```
monitor_status{monitor_name="news.bbc.co.uk",monitor_type="http",monitor_url="https://news.bbc.co.uk",monitor_hostname="null",monitor_port="null"} 1
```

then I should see

```
monitor_status{monitor_name="news.bbc.co.uk",monitor_type="http",monitor_url="https://news.bbc.co.uk",monitor_hostname="null",monitor_port="null", my_tag_1="my_value_1", my_tag_2="my_value_2"} 1
```

Does that make sense?

chakflying commented:
If that's the data structure you want, you would probably do this:
```js
for (const tag of monitor.tags) {
  this.monitorLabelValues[tag.name] = tag.value
}
```

proffalken commented:
Looks like the tags aren't part of the `monitor` object?

`console.log(monitor);` returns the following 

```
Monitor {
  beanMeta: BeanMeta {
    noCache: false,
    fetchAs: '',
    alias: '',
    via: '',
    withCondition: '',
    withConditionData: []
  },
  _id: 5,
  _name: 'news.bbc.co.uk',
  _active: 1,
  _userId: 1,
  _interval: 60,
  _url: 'https://news.bbc.co.uk',
  _type: 'http',
  _weight: 2000,
  _hostname: null,
  _port: null,
  _createdDate: '2021-11-09 15:40:17',
  _keyword: null,
  _maxretries: 0,
  _ignoreTls: 0,
  _upsideDown: 0,
  _maxredirects: 10,
  _acceptedStatuscodesJson: '["200-299"]',
  _dnsResolveType: 'A',
  _dnsResolveServer: '1.1.1.1',
  _dnsLastResult: null,
  _retryInterval: 60,
  _pushToken: null,
  _method: 'GET',
  _body: null,
  _headers: null
}
```

Saibamen commented on server/prometheus.js:
```diff
@@ -42,6 +42,9 @@ class Prometheus {
             monitor_hostname: monitor.hostname,
             monitor_port: monitor.port
         };
+        for (var t in monitor.tags) {
```
```suggestion
        };
        
        for (var t in monitor.tags) {
```

Saibamen replied:
```suggestion
        };
        for (var tag in monitor.tags) {
```

Saibamen commented on server/prometheus.js:
```diff
@@ -42,6 +42,9 @@ class Prometheus {
             monitor_port: monitor.port
         };
+        for (var t in monitor.tags) {
+            this.monitorLabelValues[k] = monitor.tags[k];
```
This should be
```suggestion
            this.monitorLabelValues[k] = monitor.tags[tag];
```
??

Saibamen reviewed (changes requested):
Please use ESLint!

You are using `var` in your code...

proffalken commented:
> Please use ESLint!
> 
> You are using `var` in your code...

Yup, I learned Javascript nearly 20 years ago and spend most of my life writing other languages.  I had no idea that `var` had been replaced.  I've added `eslint` to my codebase and fixed the syntax issues, however it doesn't fix the actual issue that the tags in `monitor.tags` don't show up as prometheus labels.

Any ideas how I get this working?

chakflying commented:
Currently we use the ORM in a very raw way, meaning that table joins have to be done manually. You can check `monitor.js:52` for how this is done. But it does seem a bit weird that we need to have another DB query everytime we want to use tags. Maybe @louislam will know if the ORM can do this automatically.

proffalken commented:
@louislam - I'm hoping to find time to come back to this shortly, is the ORM capable of doing the links or should I just go for a raw query?

proffalken commented:
@chakflying - I've just found time to come back to this.

I've added some code that does a lookup and pulls back the tags, however I can't seem to get them to be passed through to the correct dict.

In the logs, I get the following:

```
Getting Tags for Prometheus
TAGS: [object Promise]
Found the following tags for 1 :
[
  {
    id: 1,
    monitor_id: 1,
    tag_id: 1,
    value: 'garage',
    name: 'location',
    color: '#4B5563'
  },
  {
    id: 2,
    monitor_id: 1,
    tag_id: 2,
    value: 'se-1',
    name: 'region',
    color: '#DC2626'
  }
]
```
but I'd expect to see all of the tags listed on the line that starts `TAGS`, rather than a promise?

I've tried to add an `await` to the call to `get_tags(monitor)`, but this results in the following error:

```
/home/mmw/Projects/uptime-kuma/server/prometheus.js:58
        let tags = await this.get_tags(monitor);
                   ^^^^^

SyntaxError: await is only valid in async functions and the top level bodies of modules
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1026:15)
    at Module._compile (node:internal/modules/cjs/loader:1061:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1151:10)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:999:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/home/mmw/Projects/uptime-kuma/server/model/monitor.js:8:24)
    at Module._compile (node:internal/modules/cjs/loader:1097:14)
```
Any idea how I can get around this?

proffalken commented:
OK, the latest code now takes *specific* tags and applies them as labels to Prometheus metrics.

This isn't ideal, because I'd really like to have the labels added dynamically, but it's a constraint from the upstream Prom-client package, so it will have to do for now.

![image](https://user-images.githubusercontent.com/262545/154907750-22d7e1ab-4cd7-49ef-acc2-ff133fc47844.png)

![image](https://user-images.githubusercontent.com/262545/154907925-17fe8dd0-0431-46e8-bc65-50e3b811bfb2.png)

proffalken commented:
Tags that will automatically be turned into labels are:

```
    "location",
    "region",
    "datacenter",
    "cloud_provider",
    "az",
    "rack",
    "shelf",
    "room",
    "floor"
```

and new ones can be added by updating https://github.com/louislam/uptime-kuma/pull/898/files#diff-a2ea08464c146b6af2888bdf744cb4789cac4bcda6731289d02431bcc9451363R10-R18

Once this is merged, I'll update the wiki accordingly.

Tags that do not meet one of the above names are ignored silently

Saibamen commented on server/prometheus.js:
```diff
@@ -34,6 +35,17 @@ const monitor_status = new PrometheusClient.Gauge({
+
+        const tags = await R.getAll("SELECT mt.*, tag.name, tag.color FROM monitor_tag mt JOIN tag ON mt.tag_id = tag.id WHERE mt.monitor_id = ?", [monitor.id]);
+
+        console.log("Found the following tags for " + monitor.id +" :");
+        console.log(tags);
```
```suggestion
        console.log("Found the following tags for " + monitor.id +":", tags);
```

Saibamen commented on server/prometheus.js:
```diff
@@ -55,14 +64,16 @@ class Prometheus {
             console.log("New tag created: {" + name + ": " + value + "}");
             this.monitorLabelValues[name] = value;
         }
+        }
```
Fix formatting

Saibamen commented on server/prometheus.js:
```diff
@@ -7,6 +7,15 @@ const commonLabels = [
+    "rack",
+    "shelf",
+    "room",
+    "floor"
```
`"floor",`

proffalken replied:
ah, is this another JS rule I need to be aware of? I'd usually leave off the trailing comma.

Thanks for the comments, I'll address them shortly! :)

proffalken commented:
@chakflying / @louislam - any chance you can take a look at this please?

louislam commented:
Looking for a tester who are using prometheus to test this pull request.

Computroniks commented:
# Intermediary test results

Here are some intermediary results from my testing so far. Please note that I haven't quite finished testing the PR yet so I will provide some more updates probably tomorrow.

## Test environment

### Server
* OS: Debain 11
* Kernel: Linux 5.10.0-11-amd64
* Platform: VMware
* Architecture: x86-64
* `node -v`: v17.4.0
* `pm2 -v`: 5.1.2
* Previously ran Uptime Kuma 1.13.1 in production using pm2
### Client
* OS: Kali GNU/Linux Rolling 2021.1
* Kernel: Linux 5.10.0-kali7-amd64
* Platform: Acer laptop
* Architecture: x86-64
* Browser: 99.0.4844.82 (Official Build) (64-bit) 
* JavaScript: V8 9.9.115.9
* User agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36

## Results

Hmm, I seem to be getting some slightly odd results. There are a couple of things that seem a touch unusual. When I add a tag to a monitor, the prometheus metric gets repeated, one with the tag included and one without.
```
monitor_cert_days_remaining{monitor_name="google.com",monitor_type="http",monitor_url="https://google.com",monitor_hostname="null",monitor_port="null"} 61
monitor_cert_days_remaining{monitor_name="google.com",monitor_type="http",monitor_url="https://google.com",monitor_hostname="null",monitor_port="null",location="google"} 61
```

Another thing I noticed, when I create a new monitor with a tag, it doesn't appear in the prometheus metrics and only appears if I edit the monitor and resave it.

Steps to reproduce
* Create a new monitor and assign a tag from the list in a previous comment
* Save the monitor
* Visit the /metrics page and note that the new monitor does not contain the tag
* Edit the monitor and resave it
* Visit /metrics again and note that the previous metric without the tag is still present but a new one with the tag has also been added.

Note that these have all been tested with a database that previously came from 1.13.1 before issuing the following commands:

```
gh pr checkout 898
npm install
npm run build
pm2 restart
``` 

I will try again with a clean database and clean `node_modules` directory.

Computroniks commented:
> Hmm, I seem to be getting some slightly odd results. There are a couple of things that seem a touch unusual. When I add a tag to a monitor, the prometheus metric gets repeated, one with the tag included and one without.
> 
> ```
> monitor_cert_days_remaining{monitor_name="google.com",monitor_type="http",monitor_url="https://google.com",monitor_hostname="null",monitor_port="null"} 61
> monitor_cert_days_remaining{monitor_name="google.com",monitor_type="http",monitor_url="https://google.com",monitor_hostname="null",monitor_port="null",location="google"} 61
> ```
> 
> Another thing I noticed, when I create a new monitor with a tag, it doesn't appear in the prometheus metrics and only appears if I edit the monitor and resave it.

This does seem to persist after I deleted the database and dist folders as well as `node_modules` and stoped the pm2 service and then ran the following.
```
npm install
npm run build
npm start
```

Although when the service is restarted the duplication disappears. I presume that this is not so much an issue with the PR but more an issue with the way the Prometheus metrics page is constructed

Computroniks reviewed (changes requested):
Aside from the issue with the repetition of metrics and metrics not being created on monitor creation, this PR seems to work. It links in with Grafana perfectly and allows the user to group their graphs with ease. I also tested using emojis for tag values and this posed no problem. The restriction on the tags isn't much of an issue as the ones you have allow for great flexibility. 

The graph below illustrates the issue of repeated metrics on settings change.
![image](https://user-images.githubusercontent.com/67638596/163065517-9670396d-de14-46e0-97cd-5eb7b023da40.png)

The prometheus query being run is `monitor_response_time{location="google", monitor_name="translate"}`

The yellow line was the original metric for translate.google.com. We can see that it responds normally. The settings for this monitor were then updated at 23:23 (start of green line), at this point the green line takes over the expected behavoir and the yellow line just stays at it's previous value. When Uptime Kuma is restarted, this would disappear and only one line would be shown.

Great work on this PR so far! 👍

chakflying commented:
Should probably rebase on master first, #1136 should have fixed the duplicate metrics issue, but I'm guessing it still doesn't handle editing tags correctly.

Computroniks commented:
@chakflying I tried it again after rebasing but as you suspected, it doesn't seem to handle the changes in tags and the issue with the tag only appearing after second save is still present.

proffalken commented:
Wow, thanks, I'd completely missed the updates on this for some reason, great to see it making progress!

proffalken commented:
Apologies, life got in the way, this has now been rebased on master, hopefully it can be merged soon?

Saibamen commented:
Please fix ESLint warnings

proffalken commented:
@Saibamen - this is now done.  If we're going to enforce the fixing of warnings as well, surely we should update the CI so it fails on that too?

I'd not bothered to check the logs because the tests were green, so I missed that there were lint warnings too. :(

proffalken commented:
@Computroniks - I'd appreciate if you could test this again for me, especially with the issue you're seeing around tags etc. as I'm not quite sure I've followed that part of the conversation.

Computroniks commented:
Sure, I will take another look tonight

proffalken commented:
> > I checked the source code. Since `addMonitorTag` is a separated call which I guess it called after `addMonitor`. That means when the monitor saved and restarted internally, at this moment, those tags haven't added yet.
> > https://github.com/louislam/uptime-kuma/blob/4df147786da8835d4260dde115937935bd643df1/server/server.js#L928-L930
> 
> I think this issue is not fixed yet?
> 
> Recap: The tag(s) is/are not active for the first save.

Oh, ok, thanks, this is the bit I've not quite followed.

Is there a "quick" solution to this I can implement? Should this logic be added somehow to `addMonitor` instead?

proffalken commented:
I could really do with some help here from either @Computroniks or @louislam as I'm well out of my depth!

louislam commented:
I don't exactly remember the flow of prometheus endpoint.

I think you can try to recreate the `Prometheus` object in the `addMonitorTag` event

- [/server/model/monitor.js] add a function maybe call `monitor.refreshPrometheus()`
- [/server/server.js] inside `addMonitorTag`, get the monitor (server.monitorList[monitorID]) and call `monitor.refreshPrometheus()`

proffalken commented:
Thanks, I'll find some time to give that a go

spali commented:
@proffalken 
When you get time to continue, would love to see some specific labels/tags that are often used:

- `identifier`, `hostname` or `nodename`
  Often used to join to other metrics in dashboard.
  Maybe `identifier` would be sufficient. At least that would be my preference 😉 
  Because it is open to what exactly it represents and could be relabeled in Prometheus to the specific name by everyone.
- `depends_on`
  can be used by alert manager to make rules for dependencies of services or devices like gateways.

proffalken commented:
@spali the idea was to leave the actual tags up to the end-user as labels in the UI.

Having said that, I'd be willing to consider some "standard tags", although there's a lot going on in my personal life right now and I'm unlikely to get to this any time soon.

spali commented:
> Having said that, I'd be willing to consider some "standard tags", although there's a lot going on in my personal life right now and I'm unlikely to get to this any time soon.

Even better if it works.
Sorry, then I have misunderstood a comment about the prom-client restriction.

Take your time, I know that problem.

spali commented:
@proffalken 
Would like to take over if it's ok.

@louislam 
I investigated the code a bit and see some problems with the current approach and suggested approach.
Here some points (please correct if I got something wrong).

1. I would suggest to rely on `monitor.start()` and `monitor.stop()` which is called anyway on any "Save" on the monitor (includes editing tags). So if we cache the labels/tags names in the monitor instance, we can remove the Prometheus metric instance on `stop` and create it on `start` without any database access. Would simplify and be a bit cleaner as now, but same functionality.
2. change `monitor.start()` and `monitor.stop()` to async as preparation for later database access which is async (required to get all distinct label names). So any current call to these methods would require an `await`.
3. Now to the tag to label feature....
  Due the restriction on the `prom-client` which only allows a set of labels at creation time of the whole metric, it's required to recreate the whole metric including all metric instances for all monitors. 
  Because on `Prometheus` class level we don't know the other monitors and `prom-client` does not offer to get the unique "set's" of labels back, we need to drop all metric instances and the metric completely and recreate the metric and instances for each monitor.
  The problem, we need a list of the all unique tag/label combinations with the corresponding last value used in the metric.

   3.1. We clear all metrics and just restart all monitors (I think, not an option).
   3.2. We just clear all metrics and wait till next monitor heartbeat to start adding itself to the metrics again (not nice for monitoring).
   3.3. We can hold a `static` "cache" on the `Prometheus` prototype with labels to value mapping to recreate everything.
   3.4. We can hold a `static` "cache" on the monitor level with tags to value mapping to recreate everything.
   3.5. We filter new labels out on metric updates and only build the full label list at first monitor start. Like 3.1 but rely on the user to restart the server to get new labels populated to the metrics. Edit existing labels wouldn't need a restart.
   3.6 Same as 3.5 but give the user a button to restart clear all metrics and restart all monitors (same effect as server restart).


Personally I would prefer 3.3 on the cost of memory. Or if memory and simplicity is a concern 3.5.

proffalken commented:
@spali - more than happy for you to take this on, my knowledge of NodeJS could be written on the back of a small spider and still have room for most of Shakespeare's sonnets.

Nice to see someone who has time and knows what they're doing trying to make sense of my code! :)

spali commented:
example of 3.5 solution here: https://github.com/louislam/uptime-kuma/commit/054b3ec1c5c0369bc9ab4f95b2f02d27cae96c62
Would be ready to create new PR or you can push my commit into this PR.

If you want to split the change to async methods in a dedicated PR, let me know.

burakksglu commented:
Hey everyone! Your suggestions are great. I was testing the Prometheus endpoint with Grafana and found this PR. In the endgame, the Prometheus endpoint should include some of the parameters in api.get_monitors() of API also. Even the Tag parameter alone would add much more flexibility.

For dashboarding and reporting purposes, these would be great additions for users that can't access UI because of the lack of user permissions:

'accepted_statuscodes' 'active' 'expiryNotification' 'id' 'interval' 'keyword' 'maintenance' 'method' 'name' 'port' 'proxyId' 'pushToken' 'tags' 'type' 'upsideDown' 'url'

Reference:[https://uptime-kuma-api.readthedocs.io/en/latest/api.html#uptime_kuma_api.UptimeKumaApi.get_monitors
](url)

michaelyork commented:
Any progress here by chance? Happy to sponsor someone's work to get this shipped!

spali commented:
First preparation PR was merged.
Waiting for https://github.com/louislam/uptime-kuma/pull/2830 to be merged. So I can create the last PR which should solve/replace this PR.

Wanted to split it to make the reviews easier.

mabed-fr commented:
Do you want more tests ? cela semble bien fonctionner

mateuszdrab commented:
> Any updates on this?
> 
> waiting for that for a long time...

I'd also appreciate this being merged

CommanderStorm commented:
@mabed-fr @mhkarimi1383 @mateuszdrab 

Please refrain from posting `+1` / requests for update-things on PRs, as this makes reviewing it even harder.
PRs are for **noting problems and working on them**, while for updates, there is a `subscribe to this PR`-button at the top.
[We use 👍🏻 on PRs to prioritise work](https://github.com/louislam/uptime-kuma/pulls?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

What @spali stated is still true
> First preparation PR was merged.
> Waiting for #2830 to be merged. So I can create the last PR which should solve/replace this PR

mabed-fr commented:
Where is the PR, what are we missing?

CommanderStorm commented:
#2830 was merged

mabed-fr commented:
> #2830 was merged 1 month ago => previous block is now resolved

What is the logical next step?

spali commented:
See [4704](https://github.com/louislam/uptime-kuma/pull/4704)
I didn't had time to retest after fixing the merge conflicts... wrote code over a year ago ... I would appreciate if someone could test a bit my PR.

CommanderStorm commented:
Thanks for the wor that went into this PR

Closing in favor of https://github.com/louislam/uptime-kuma/pull/4704 as said PR seems to also handle the editing of tags
